mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala

```diff
@@ -138,24 +138,59 @@ private[ml] abstract class LSHModel[T <: LSHModel[T]]
       // Limit the use of hashDist since it's controversial
       val hashDistUDF = udf((x: Seq[Vector]) => hashDistance(x, keyHash), DataTypes.DoubleType)
       val hashDistCol = hashDistUDF(col($(outputCol)))
-
-      // Compute threshold to get around k elements.
-      // To guarantee to have enough neighbors in one pass, we need (p - err) * N >= M
-      // so we pick quantile p = M / N + err
-      // M: the number of nearest neighbors; N: the number of elements in dataset
-      val relativeError = 0.05
-      val approxQuantile = numNearestNeighbors.toDouble / count + relativeError
       val modelDatasetWithDist = modelDataset.withColumn(distCol, hashDistCol)
-      if (approxQuantile >= 1) {
-        modelDatasetWithDist
+
+      val spark = dataset.sparkSession
+      import spark.implicits._
+
+      if (numNearestNeighbors < 1000) {
+        val r = Random.nextInt
+        val distColIdx = modelDatasetWithDist.schema.fieldNames.indexOf(distCol)
+        val rows = modelDatasetWithDist
+          .rdd
+          .map { row =>
+            val dist = row.getDouble(distColIdx)
+            (r, (dist, row))
+          }.aggregateByKey(new BoundedPriorityQueue[(Double, Row)](numNearestNeighbors)(
+            Ordering.by[(Double, Row), Double](_._1).reverse))(
+            seqOp = (c, v) => c += v,
+            combOp = (c1, c2) => c1 ++= c2
+          ).flatMap { case (_, c) => c.iterator.map(_._2) }
+        spark.createDataFrame(rows, modelDatasetWithDist.schema)
```

Comment:
This method extract numNearestNeighbors rows, however, it fails in `BucketedRandomProjectionLSHSuite`.

I checked the datasets, and found that existing method based on `approxQuantile`/`QuantileSummaries` will generate much more candidate items in the step:
numNearestNeighbors=100, modelSubset.size=231.

Method in current commit will only extract 100 candidates even if there are repeated values, while existing method in master will extract 231 candidates. Then the new method will fail in testsuite because its final precision is not good.

I wonder that may be we need more candidates here to generate more accuracy final result?
Should I just remove the method base on Top-K, or use a scaled threshold (numNearestNeighbors*3 or numNearestNeighbors*3) here?

Reply:
The log of both methods in testsuites:
`approxQuantile/QuantileSummaries`:
numNearestNeighbors=100, modelSubset.size=231, threshold=0.0

topK
numNearestNeighbors=100, modelSubset.size=100, threshold=0.0

Reply:
There is only one test for singleProbe=false:
numNearestNeighbors=100,
the threshold computed by previous `sort` or current `approxQuantile` is 0.0,
filtered with threshold=0.0, there are 231 items.